networkx/algorithms/community/label_propagation.py

```diff
@@ -0,0 +1,178 @@
+    Parameters
+    ----------
+    G : graph
+    An undirected NetworkX graph.      
+    
+    Returns
+    -------
+    communities : list
+    List of the communities (sets) 
+    
+  
+    References
+    ----------
+    .. [1] Cordasco, G., & Gargano, L. (2010, December). Community detection 
+       via semi-synchronous label propagation algorithms. In Business 
+       Applications of Social Network Analysis (BASNA), 2010 IEEE International 
+       Workshop on (pp. 1-8). IEEE.
+    """
+    coloring = _color_network(G)
+    labeling = _uniquely_label(G)
+    nrounds = 1
+    while not _labeling_complete(labeling, G):
+        nrounds += 1
+        _update_labels(labeling, coloring, G)
+        
+    for label in set(labeling.values()):
+        yield set((x for x in labeling if labeling[x] == label))
+    
+    
+def _break_color_tie(current, labels):
+    """Uses Prec-Max tie-breaking to break the ties, as laid out in:
+       'Community Detection via Semi-Synchronous Label Propagation Algorithms'
+       Cordasco and Gargano, 2011
+
+       If the labels set specified is empty than the current label is returned.
+
+       Specified set of labels is assummed to be one of integers.
+    """
+    if len(labels) == 0 or current in labels:
+        new_label = current
+    else:
+        new_label = max(labels)
+    return new_label
+
+def _calculate_label_frequencies(node, labeling, G):
+    """Counts up the labels of the neighbors of the specified node. Returns a
+       dictionary from the label to the frequency.
+    """
+    counts = dict()
+    for q in G.neighbors(node):
+        qlabel = labeling[q]
+        if qlabel not in counts:
+            counts[qlabel] = 1 
+        else:
+            counts[qlabel] += 1              
+    return counts
+
+def _color_network(G):
+    """Colors the network so that neighboring nodes all have distinct colors.
+       Returns a dict of set of nodes and also a lookup of nodes to colors, in 
+       a tuple in that order.
+    """
+    coloring = dict() # color => set(node)
+    lookup = dict() # node => color
+    finally_colored_nodes = set()
+    
+    for n in G.nodes():
+        if n not in lookup:
+            color = 0                 
+        else:
+            color = lookup[n]
+        lookup[n] = color
+        finally_colored_nodes.add(n)
+        if color not in coloring: 
+            coloring[color] = set()
+        coloring[color].add(n)
+        for q in G.neighbors(n):
+            if q not in finally_colored_nodes: 
+                lookup[q] = lookup[n] + 1
+    
+    return coloring
+
+def _form_communities(labeling, G):
+    """Determines the communities from the labels of the network, returning a
+       dict of sets of nodes.
+    """
+    communities = dict() # label => set(nodes)
+    for n in G.nodes():
+        label = labeling[n]
+        if label not in communities: 
+            communities[label] = set()
+        communities[label].add(n)
+    return communities
```

Comment:
Simpler:

    return {k: set(v) for k, v in groupby(sorted(G, key=labeling.get), key=labeling.get)}